Contributor Guide updates for examples & codeowner definition (#615)
fairlearn/fairlearn: docs/contributor_guide/contributing_example_notebooks.rst

   Avoids abstraction traps.
   See `Selbst et al. (2020) <https://andrewselbst.files.wordpress.com/2019/10/selbst-et-al-fairness-and-abstraction-in-sociotechnical-systems.pdf>`_.
4. **Substantiated**: Discusses trade-offs and compares alternatives.

>    Describes why using particular Fairlearn functionalities makes sense if

Reviewer: This sentence is a bit tautological.

Reviewer 2: Maybe something like - "Describes why using particular Fairlearn functionalities make sense for this task or use case." ?

Author: But we may not be using Fairlearn in every use case (?)

   Fairlearn is used.
5. **For developers**: Speaks the language of developers and data scientists.
   Considers real practitioner needs.

>    Fits within the lifecycle of real practitioner work.

Reviewer 3: examples are sometimes there just to show how somebody could use the code, they don't have to be a real usecase. Sometimes showing how one can use a part of the API doesn't need to show why one needs to use that part of the API.

Author: I guess that's that leads to the bigger question of whether it's responsible to show someone how to use something without explaining when/how and when/how not to use it at the same time. If something is really just about "how one can use a part of the API" that's what the user guide (or perhaps even API doc) is for, isn't it?

Reviewer 3: To me the user guides can go more into detail than examples. Examples are mostly about the code, whereas the user guide gives context and talks about stuff.

Author: That's not really how we've been talking about examples as you can tell from that entire file. I'm not really writing this from scratch here either, it's been there for a while and I'm just making updates. The way we've talked about examples for Fairlearn as a "special" kind of open source project with a sociotechnical focus is that examples are mainly for showing practitioners how to use this properly/responsibly rather than "just code".

Let's loop in @mmadaio @MiroDudik @riedgar-ms @hildeweerts since this is a rather big question. Wouldn't want that to get lost.

Reviewer 4: Yeah... We had a long conversation about this a while back. At the time we concluded that example notebooks should be more like use cases rather than short code samples. We could revisit this question. But rather than discussing these requirements hypothetically, I'd be in favor of doing it in context of a concrete example that would fall into such a "short sample code" category, but fail our current notebook requirements?

My current thinking around where different kinds of examples fit into our documentation:
1. **API docs** [.py]: The docs describe technical functionality in isolation, and so I think that they are the best place for short examples of functionality. However, they might not be the right place for examples that deal with several parts of Fairlearn (e.g., compare several algorithms).
2. **User guide** [.rst]: I think about this basically as a tutorial, almost a textbook. Here it should be pretty straightforward to provide context and also have sections that can be more technically focused. Unlike API docs, this might be a good place to present examples that mix different types of functionality.
3. **Example notebooks** [.ipynb] (note that we haven't been calling these "Examples"): They are basically use cases and what we're currently describing in [Contributing example notebooks](https://fairlearn.github.io/v0.5.0/contributor_guide/contributing_example_notebooks.html).

Unfortunately, `.ipynb` is not GitHub friendly, and so example notebooks are `.py`.

Reviewer 5: I like @MiroDudik's thinking. I'd suggest to be very explicit in naming and call examples in (1) "code examples" and also add a link to the relevant section in the user guide (this would only work if we're giving every object/function, but I think that's something we want to work on anyways, right?).

Author: Yes! I'll rewrite this PR based on today's discussion in the community call :-) Stay tuned. I'll make sure to ping you all when I've worked in the changes.

   See `Holstein et al (2019) <https://arxiv.org/pdf/1812.05239.pdf>`_,
   `Madaio et al. (2020) <http://www.jennwv.com/papers/checklists.pdf>`_.

Please keep these in mind when creating, discussing, and critiquing examples.

Process
^^^^^^^

All current efforts are tracked through items in the corresponding
`Project Board <https://github.com/fairlearn/fairlearn/projects/3>`_.
If you'd like to suggest a different kind of project, please
`open a new issue <https://github.com/fairlearn/fairlearn/issues/new/choose>`_
if you cannot find a similar one that is already tracked in the project board.
If you'd like to collaborate with others from the community, please

> :ref:`reach out <communication>` to share your idea.

Reviewer 3: Do you intend to have ALL issues on the project board? I don't think that'd be sustainable in the long run. Users can simply search in the issue tracker, the board is more like what we're focusing on that a given time?

Author: Hmm, so far that was indeed my plan. I guess 50 issues are a lot easier to handle than 1500 (scikit-learn). It's sort of nice to have similar types of tasks grouped together, which can be done with tags, too. We should discuss this further in #504 . I'm really open to it either way. I'll update this text to say "if you cannot find a similar one in the issue tracker".

In the issue, make sure to outline your goals, describe your audience, and
provide a high-level overview of what you would like to show.
Note that the examples do not need to contain code.
In some application contexts it may be preferable not to build a
technical system, so a textual explanation is perfectly acceptable.

> All contributions should abide by the guidelines outlined above, though.

Reviewer 3: I'm glad you're volunteering for this :D if fairlearn starts getting adopted, soon we may not have the resources to answer those questions.

Author: Oh definitely! That's a problem I'd love to solve if/when we get to it. Also, we can expand to have more maintainers if we really have such an active community.

Once the issue is on GitHub, members from the community can respond with

> questions, comments, and perhaps express their interest in joining the
> project. Once the preliminary questions are sorted out, the participants of
> the project can

Reviewer 3: ```suggestion
questions, comments, and perhaps express their interest in fixing the
issue. Once the preliminary questions are sorted out, somebody
can
```

people sometimes just come by and fix an issue and leave. They don't become "a part of the project" necessarily, and that's okay.

Author: Used your suggestions (manually to avoid offending DCO...)

`open a pull request <https://github.com/fairlearn/fairlearn/compare>`_.
Everyone from the community is encouraged to review the pull request and
provide feedback. The acceptance process is identical to all other
contributions as outlined :ref:`here <development_process>`, except that

> the additional criteria at the top of this page apply additionally.

Reviewer 3: maybe have a reference here to be able to follow and see what that criteria was?

Reviewer: Also 'additional criteria apply additionally' ?

Whether and how project members communicate is up to the members themselves.
This could include any of the following:

- a dedicated channel on [Gitter](https://gitter.im/fairlearn) in the
  Fairlearn community
- a Slack channel or Teams team
- email

> - recurring or ad-hoc meetings

Reviewer 3: I think we should encourage asynchronous communication on github, with the added possibility of whatever public chat channel we have. Otherwise people from certain timezones and people who are not involved in the discussion are excluded easily.

Author: Indeed. I think the "Discussions" might actually lend itself to this (more so that issues since they are linear). Let me update that. For this particular point we really want the individual groups to make those decisions rather than us forcing anything onto people. If there are a few people who want to have this particular discussion over email they can certainly do that.

Author: Updated this to encourage people to use the more inclusive async communication. Thanks for suggesting!

Reviewer 3: I would still argue that this should be strongly discourages since it excludes everybody who's not on that email thread.

If the project group would like to discuss their project with the community
we can use one of our
:ref:`community calls <community_calls>` for this
purpose.

> If you have problems with any parts of this process or feedback please
> :ref:`reach out <communication>`.

Reviewer 3: ```suggestion
If you think we can improve any parts of this process or feedback please
:ref:`reach out <communication>`.
```

Author: incorporated manually (because of DCO)
